### ISSUE TITLE
Annotate most functions with #[must_use] and add Drop impl for ptr and handles?

### ISSUE THREAD

**Issue by JasonKleban** [retep998/winapi-rs#298]
Would it be good and appropriate to annotate most functions in this crate with `#[must_use]` and add api-specific Drop trait implementations so that the allocated or reserved items get clean up properly?  Is this practical?  Would it actually make this crate more safe?  Would it make it fully safe and without an runtime performance impact?

```
#[must_use]
pub fn CreateWindowExW(
        dwExStyle: DWORD, lpClassName: LPCWSTR, lpWindowName: LPCWSTR, dwStyle: DWORD, x: c_int,
        y: c_int, nWidth: c_int, nHeight: c_int, hWndParent: HWND, hMenu: HMENU,
        hInstance: HINSTANCE, lpParam: LPVOID,
    ) -> USER32_HWND_DestroyWindow;
```

`USER32_HWND_DestroyWindow` is a little clunky - but it indicates that it would have to be defined specifically for this library, since as I understand it each library exposes its own appropriate `free`/`release` functions and even that within a library for a particular type there might be different ones or different expectations made by the library.

But that wouldn't address the need to supply the return value of this function to some other function that doesn't care about the library that the value came from.  It seems like values that the function takes still needs to be the broader type without the ownership annotations and that values it returns (or mutates) would need the ownership annotations.  I'm not sure if/how Rust could express it.

I also could be talking nonsense and if so, can somebody see where I lost it and help me get it?

**Comment by tomaka** [comment]
Winapi is not a high-level crate.
If you look at libc, libgit-sys, libopenal-sys, etc., none of them provide utilities.

**Comment by Eh2406** [comment]
Hi,

You got it. That all makes sense! But this library has intentionally chosen not to do that so as to stick just to wrapping the WinAPI. This means that if one wants a more rusty interface one can wrap this library in another library. I think, several attempts have been made to make a wrapping library. Some like [clipboard-win](https://crates.io/crates/clipboard-win) are focused, compleat, and used. Others, are no longer maintained.

TLDR: That is a great library{s} and should be built on top of this one.

**Comment by retep998** [comment]
Any sort of high level wrapper is definitely out of scope for this crate. The highest level thing I do in winapi is the `RIDL!` stuff for COM interfaces. However do feel free to write your own libraries with higher level wrappers around the primitives provided here.

**Comment by JasonKleban** [comment]
Ok, thanks for the explanation.

Does such wrapping incur a run-time penalty?  If so, is it avoidable.  If not, why is that out of scope for this crate or any crate?  Is there a benefit to keeping the low and high-level crates separate?

Are there conventions used in the winapi that are too dynamic to encode in Rust's type system?

**Comment by Eh2406** [comment]
> Does such wrapping incur a run-time penalty?
> Are there conventions used in the winapi that are too dynamic to encode in Rust's type system?

Generally no, but I am new to both. :-P

>  Why is that out of scope for this crate or any crate?

Wrapping WinAPI at a low level is a massive product. @retep998 does an enormous amount of work to keep it consistent, correct, and progressing. Scope kreep is a great way to develop burnout. 

> Is there a benefit to keeping the low and high-level crates separate?

More generally, yes.

Let's say that I don't {like | can't use} a high-level crates for some reason. I nead to reimplement the low level so I can build my abstractions. If the low-level is a separate crate than we can share that part. The more code gets reused, and the more people to look at it, the more bugs are fixed.

Let's say I am developing in conjunction with legacy code (possibly over ffi) and the rest of my code base does not know about rusts abstractions. It may be easier to work with the same data structures as the legacy code as provided in the low-level crates.

**Comment by retep998** [comment]
>Does such wrapping incur a run-time penalty?

Only if you have optimizations disabled. It does however incur a compile time penalty.

>Are there conventions used in the winapi that are too dynamic to encode in Rust's type system?

COM interfaces are probably the most complex thing to encode in Rust's type system, due to inheritance. Really the bulk of the problem isn't the type system, but rather enforcing various invariants without sacrificing ease of use. Asynchronous operations, shared ownership with the OS, buffers, threading, there's a lot of details that make designing a good API not entirely straightforward, so it is better to allow people to experiment and create all sorts of high level wrappers of their own.
